In [0]:
%run ../Includes/Copy-Datasets

Lets take a look at the delta logs for the bronze table 

In [0]:
files = dbutils.fs.ls("dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log")
display(files)

path,name,size,modificationTime
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000000.crc,00000000000000000000.crc,2381,1730698792000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000000.json,00000000000000000000.json,1414,1730698790000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000001.crc,00000000000000000001.crc,5977,1730698797000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000001.json,00000000000000000001.json,4276,1730698796000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000002.crc,00000000000000000002.crc,10169,1730698809000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000002.json,00000000000000000002.json,4979,1730698808000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000003.crc,00000000000000000003.crc,16475,1731203210000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000003.json,00000000000000000003.json,7119,1731203204000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000004.crc,00000000000000000004.crc,19979,1731203234000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000004.json,00000000000000000004.json,4289,1731203231000


Lets take a look at one of the transaction log files, and look at the statistics 

In [0]:
display(spark.read.json("dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000001.json"))

add,commitInfo,txn
null,"List(0930-040814-iga72s2f, Databricks-Runtime/13.3.x-scala2.12, true, WriteSerializable, List(2162048479376534), STREAMING UPDATE, List(5, 59717, 812, 0), List(0, Append, 4b35ec04-0dae-4bac-bd06-ed3a48c6cf12), 0, List(false), 1730698795278, e4d743cc-bced-4c9a-8e9c-d7d16674fbef, 8165668582725249, franyang@deloitte.com.au)",null
null,null,"List(4b35ec04-0dae-4bac-bd06-ed3a48c6cf12, 1730698795277, 0)"
"List(true, 1730698795000, List(books, 2021-11), topic=books/year_month=2021-11/part-00000-22e93f57-988c-4da8-9caa-3800f13eeb26.c000.snappy.parquet, 2922, {""numRecords"":12,""minValues"":{""partition"":0,""offset"":18274,""timestamp"":""2021-11-07T17:16:33.000Z""},""maxValues"":{""partition"":4,""offset"":18285,""timestamp"":""2021-11-10T16:41:31.000Z""},""nullCount"":{""key"":0,""value"":0,""partition"":0,""offset"":0,""timestamp"":0}}, List(1730698795000000, 1730698795000000, 1730698795000000, 268435456))",null,null
"List(true, 1730698795000, List(customers, 2021-11), topic=customers/year_month=2021-11/part-00000-7d5c37fe-cfc2-4f69-a9bd-ca1c41f8491d.c000.snappy.parquet, 11865, {""numRecords"":105,""minValues"":{""partition"":0,""offset"":618727,""timestamp"":""2021-11-12T01:14:00.000Z""},""maxValues"":{""partition"":4,""offset"":618831,""timestamp"":""2021-11-30T22:29:00.000Z""},""nullCount"":{""key"":0,""value"":0,""partition"":0,""offset"":0,""timestamp"":0}}, List(1730698795000001, 1730698795000001, 1730698795000001, 268435456))",null,null
"List(true, 1730698795000, List(customers, 2021-12), topic=customers/year_month=2021-12/part-00000-570bff98-edde-4e86-b233-ca5dec39bfe4.c000.snappy.parquet, 18927, {""numRecords"":195,""minValues"":{""partition"":0,""offset"":618832,""timestamp"":""2021-12-01T08:07:00.000Z""},""maxValues"":{""partition"":4,""offset"":619026,""timestamp"":""2021-12-31T15:09:00.000Z""},""nullCount"":{""key"":0,""value"":0,""partition"":0,""offset"":0,""timestamp"":0}}, List(1730698795000002, 1730698795000002, 1730698795000002, 268435456))",null,null
"List(true, 1730698795000, List(orders, 2021-11), topic=orders/year_month=2021-11/part-00000-2e92aae8-8e6d-4a4a-ba9d-7f7421077db4.c000.snappy.parquet, 11281, {""numRecords"":208,""minValues"":{""partition"":0,""offset"":39127899,""timestamp"":""2021-11-12T09:15:00.000Z""},""maxValues"":{""partition"":4,""offset"":39128064,""timestamp"":""2021-11-30T22:12:00.000Z""},""nullCount"":{""key"":0,""value"":0,""partition"":0,""offset"":0,""timestamp"":0}}, List(1730698795000003, 1730698795000003, 1730698795000003, 268435456))",null,null
"List(true, 1730698796000, List(orders, 2021-12), topic=orders/year_month=2021-12/part-00000-7000e2e6-6065-4576-be25-1f95997e1b2c.c000.snappy.parquet, 14722, {""numRecords"":292,""minValues"":{""partition"":0,""offset"":39128065,""timestamp"":""2021-12-01T00:07:00.000Z""},""maxValues"":{""partition"":4,""offset"":39128298,""timestamp"":""2021-12-20T09:14:00.000Z""},""nullCount"":{""key"":0,""value"":0,""partition"":0,""offset"":0,""timestamp"":0}}, List(1730698795000004, 1730698795000004, 1730698795000004, 268435456))",null,null


This is really useful because it you are filtering for an offset that has a greater offset than that of the maximum value in, delta lake will just skip this file completely. If you look at row 4, and look at the summary statistics, we see that the maximum value for offset is 618831, so if we are querying, and we are looking for an offset greater than 618831, then this file will be skipped completely because the offset we are querying for is less than the maximum offset for this data file 

In [0]:
%sql
SELECT COUNT(*) FROM bronze

count(1)
8036


Lets take a look at the transaction log checkpoints - the parquet files that are created every 10 json files 

In [0]:
files = dbutils.fs.ls("dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log")
display(files)

path,name,size,modificationTime
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000000.crc,00000000000000000000.crc,2381,1730698792000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000000.json,00000000000000000000.json,1414,1730698790000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000001.crc,00000000000000000001.crc,5977,1730698797000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000001.json,00000000000000000001.json,4276,1730698796000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000002.crc,00000000000000000002.crc,10169,1730698809000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000002.json,00000000000000000002.json,4979,1730698808000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000003.crc,00000000000000000003.crc,16475,1731203210000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000003.json,00000000000000000003.json,7119,1731203204000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000004.crc,00000000000000000004.crc,19979,1731203234000
dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000004.json,00000000000000000004.json,4289,1731203231000


In [0]:
display(spark.read.parquet("dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000010.checkpoint.parquet"))

---------------------------------------------------------------------------
AnalysisException                         Traceback (most recent call last)
File <command-7401517899770341>, line 1
----> 1 display(spark.read.parquet("dbfs:/user/hive/warehouse/bookstore_eng_pro.db/bronze/_delta_log/00000000000000000010.checkpoint.parquet"))

File /databricks/spark/python/pyspark/instrumentation_utils.py:48, in _wrap_function.<locals>.wrapper(*args, **kwargs)
     46 start = time.perf_counter()
     47 try:
---> 48     res = func(*args, **kwargs)
     49     logger.log_success(
     50         module_name, class_name, function_name, time.perf_counter() - start, signature
     51     )
     52     return res

File /databricks/spark/python/pyspark/sql/readwriter.py:544, in DataFrameReader.parquet(self, *paths, **options)
    533 int96RebaseMode = options.get("int96RebaseMode", None)
    534 self._set_opts(
    535     mergeSchema=mergeSchema,
    536     pathGlobFilter=pathGlobFilter,
   (...)
 